lib/puppet/resource_api/property.rb

```diff
@@ -69,10 +83,25 @@ def rs_value
   # method overloaded only for the :ensure property, add option to check if the
   # rs_value matches is. Only if the class is child of
   # Puppet::ResourceApi::Property.
-  def def_insync?
+  def def_ensure_insync?
     define_singleton_method(:insync?) { |is| rs_value.to_s == is.to_s }
   end
 
+  def def_custom_insync?
+    define_singleton_method(:insync?) do |is|
+      provider    = @referrable_type.my_provider
+      context     = @referrable_type.context
+      should_hash = @resource.rsapi_canonicalized_target_state
+      is_hash     = @resource.rsapi_current_state
+      title       = @resource.rsapi_title
+
+      provider_insync_result = provider.insync?(context, title, @attribute_name, is_hash, should_hash) if provider.respond_to?(:insync?)
+      return provider_insync_result unless provider_insync_result.nil?
```

Comment:
I'm sorely tempted to require that this should require only one of `[true, false, nil]` to be returned to avoid accidental leakage of values from ruby's return-less returns.

Either way, the resource api should drop any reference to provider-specified data here and cast explicitly to bool: `return !!provider_insync_result unless ...`